### PR TITLE
76 add pagination for stripe orders on account page

### LINKED ISSUE #76: Add pagination for stripe orders on account page
https://github.com/jcreek/SvelteKitSaasBoilerplate/pull/69#discussion_r1826546931

### PR DESCRIPTION
**Pull request by jcreek**
Closes #76 
image

image

## Summary by CodeRabbit

- **New Features**
	- Introduced pagination for user transactions, allowing users to navigate through transaction pages.
	- Added a new pagination form with "Previous" and "Next" buttons for easier navigation.
	- Enhanced transaction fetching to include additional parameters for improved user experience.

- **Bug Fixes**
	- Enhanced error handling for transaction fetching and pagination actions.

- **Documentation**
	- Updated user interface to reflect changes in transaction handling and pagination features.

### CODE REVIEW

File: src/routes/account/+page.svelte [hunk 1]
```diff
@@ -39,6 +39,22 @@
...
+	const handlePaginationSubmit: SubmitFunction = () => {
+		loading = true;
+
+		return async ({ result }) => {
+			transactions = JSON.parse(result.data.transactions);
+			hasNextPage = result.data.hasNextPage;
+			hasPreviousPage = !result.data.isFirstPage;
+			loading = false;
+		};
+	};
```
[the remark this answers is not in the document]
Comment (jcreek) [reply src/routes/account/+page.svelte:57]: This is unnecessary for MVP

File: src/lib/utils/supabase/admin.ts [hunk 2]
```diff
@@ -610,10 +615,21 @@ const getUserTransactions = async (userId: string) => {
...
+			customer: customerId,
+			limit: perPage
+		};
+
+		if (startingAfter) params.starting_after = startingAfter;
+		if (endingBefore) params.ending_before = endingBefore;
+
+		const charges = await stripeClient.charges.list(params);
+
+		const hasNextPage = charges.has_more;
```
[the remark this answers is not in the document]
Comment (jcreek) [reply src/lib/utils/supabase/admin.ts:628]: Unnecessary for MVP

File: src/routes/account/+page.svelte [hunk 3]
```diff
@@ -5,8 +5,8 @@
 	export let data;
 	export let form;
 
-	let { session, user, subscriptions, transactions } = data;
-	$: ({ session, supabase, user } = data);
+	let { session, user, subscriptions, transactions, pageSize } = data;
+	$: ({ session, supabase, user, transactions } = data);
```
[the remark this answers is not in the document]
Comment (jcreek) [reply src/routes/account/+page.svelte]: Done

File: src/routes/account/+page.server.ts [hunk 4]
```diff
@@ -78,5 +92,41 @@ export const actions: Actions = {
...
+		try {
+			const { transactions, hasNextPage } = await getUserTransactions(
+				session.user.id,
+				pageSize,
+				startingAfter,
+				endingBefore
+			);
+
+			const isFirstPage = transactions[0].id === firstTransactionId;
+
```
[the remark this answers is not in the document]
Comment (jcreek) [reply src/routes/account/+page.server.ts:121]: Done

File: src/routes/account/+page.server.ts [hunk 5]
```diff
@@ -19,9 +23,19 @@ export const load: PageServerLoad = async ({ locals: { supabase, safeGetSession
 
 	const subscriptions = await getUserSubscriptions(session.user.id);
 
-	const transactions = await getUserTransactions(session.user.id);
+	const transactionsData = await getUserTransactions(session.user.id, transactionsPerPage);
+
+	const transactions = transactionsData.transactions;
+	firstTransactionId = transactions[0]?.id;
+	hasMoreThanOnePage = transactionsData.hasNextPage;
```
[the remark this answers is not in the document]
Comment (jcreek) [reply src/routes/account/+page.server.ts:30]: Done

File: src/routes/account/+page.server.ts [hunk 6]
```diff
@@ -4,6 +4,10 @@ import logger from '$lib/utils/logger/logger';
 import { getUserSubscriptions, getUserTransactions } from '$lib/utils/supabase/admin';
 import { requestAccountDeletion } from '$lib/utils/supabase/admin';
 
+const transactionsPerPage = 5;
+let firstTransactionId: string;
+let hasMoreThanOnePage = false;
```
[the remark this answers is not in the document]
Comment (jcreek) [reply src/routes/account/+page.server.ts:9]: Done

**Comment by jcreek** [comment]
@OllyNicholass this PR is now ready for human review and has no remaining open coderabbit comments as all have been actioned or are irrelevant :)

**Review by OllyNicholass — APPROVED** [review]

**Comment by OllyNicholass** [comment]
Nice work bud

**Pull request merged**